pennylane_pq/devices.py

```diff
@@ -309,41 +309,44 @@ def reset(self):
         self._eng = pq.MainEngine(backend, verbose=self._kwargs['verbose'])
         super().reset()
 
-    def pre_expval(self):
-        """Flush the device before retrieving expectation values.
+    def pre_measure(self):
+        """Flush the device before retrieving observable measurements.
         """
         self._eng.flush(deallocate_qubits=False)
 
-    def expval(self, expectation, wires, par):
-        """Retrieve the requested expectation value.
+    def expval(self, observable, wires, par):
+        """Retrieve the requested observable expectation value.
         """
-        if expectation == 'PauliX' or expectation == 'PauliY' or expectation == 'PauliZ':
+        if observable == 'PauliX' or observable == 'PauliY' or observable == 'PauliZ':
             expval = self._eng.backend.get_expectation_value(
-                pq.ops.QubitOperator(str(expectation)[-1]+'0'),
+                pq.ops.QubitOperator(str(observable)[-1]+'0'),
                 [self._reg[wires[0]]])
-            # variance = 1 - expval**2
-        elif expectation == 'Hadamard':
+        elif observable == 'Hadamard':
             expval = self._eng.backend.get_expectation_value(
                 1/np.sqrt(2)*pq.ops.QubitOperator('X0')+1/np.sqrt(2)*pq.ops.QubitOperator('Z0'),
                 [self._reg[wires[0]]])
-            # variance = 1 - expval**2
-        elif expectation == 'Identity':
+        elif observable == 'Identity':
             expval = 1
-            # variance = 1 - expval**2
-        # elif expectation == 'AllPauliZ':
+        # elif observable == 'AllPauliZ':
         #     expval = [self._eng.backend.get_expectation_value(
         #         pq.ops.QubitOperator("Z"+'0'), [qubit])
         #                for qubit in self._reg]
-        #     variance = [1 - e**2 for e in expval]
 
-        if self.shots != 0 and expectation != 'Identity':
+        if self.shots != 0 and observable != 'Identity':
```

Comment:
starting to think that `shots` should be passed as `qml.expval(PauliZ(...), shots=1000)` or at evaluation `circuit(args, shots=500)` rather than at the device level :thinking:

Reply:
I don't think it would work to have it at the operation level (it becomes too fine grained, and we won't be able to support measurements with different number of shots).

The evaluation step makes most sense to me.

Reply:
True, but if we get to a stage where we're already thinking about supporting non-commuting measurements in the same circuit (by just evaluating independently), then having different numbers of shots fits that pattern.

But for the moment, I agree, that it probably makes most sense at evaluation time (so users can easily change on the fly)